In [1]:
import spacepy.pycdf as pycdf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
L1_AUX = 'cdf_data/20250621/AL1_ASW91_L1_AUX_20250621_UNP_9999_999999_V01.cdf'
L1_TH1 = "cdf_data/20250621/AL1_ASW91_L1_TH1_20250621_UNP_9999_999999_V01.cdf"
L1_TH2 = 'cdf_data/20250621/AL1_ASW91_L1_TH2_20250621_UNP_9999_999999_V01.cdf'
L2_BLK = 'cdf_data/20250621/AL1_ASW91_L2_BLK_20250621_UNP_9999_999999_V02.cdf'
L2_TH1 = 'cdf_data/20250621/AL1_ASW91_L2_TH1_20250621_UNP_9999_999999_V02.cdf'
L2_TH2 = 'cdf_data/20250621/AL1_ASW91_L2_TH2_20250621_UNP_9999_999999_V02.cdf'

In [2]:
pycdf.CDF(L2_TH2)

<CDF:
energy_center_mod: CDF_DOUBLE [17275, 50]
energy_uncer: CDF_DOUBLE [] NRV
epoch_for_cdf_mod: CDF_EPOCH [17275]
flux_uncer: CDF_DOUBLE [] NRV
integrated_flux_mod: CDF_DOUBLE [17275, 50]
integrated_flux_s15_mod: CDF_DOUBLE [17275, 50]
integrated_flux_s16_mod: CDF_DOUBLE [17275, 50]
integrated_flux_s17_mod: CDF_DOUBLE [17275, 50]
integrated_flux_s18_mod: CDF_DOUBLE [17275, 50]
integrated_flux_s19_mod: CDF_DOUBLE [17275, 50]
spacecraft_xpos: CDF_DOUBLE [17275]
spacecraft_ypos: CDF_DOUBLE [17275]
spacecraft_zpos: CDF_DOUBLE [17275]
sun_angle_tha2: CDF_DOUBLE [17275, 32, 3]
>

In [8]:
from src.preprocessing_cdf_file import *
dataset_df_01 = prepare_data(L1_AUX)

In [9]:
dataset_df_01.shape

(172786, 18)

In [10]:
dataset_df_01

epoch_for_cdf  trig_counts  coin_trig_counts  \
0      2025-06-21 00:00:06.241            0                 0   
1      2025-06-21 00:00:06.741            1                 1   
2      2025-06-21 00:00:07.241            1                 1   
3      2025-06-21 00:00:07.741            0                 0   
4      2025-06-21 00:00:08.241            0                 0   
...                        ...          ...               ...   
172781 2025-06-21 23:59:57.527            1                 1   
172782 2025-06-21 23:59:58.027            0                 0   
172783 2025-06-21 23:59:58.527            1                 1   
172784 2025-06-21 23:59:59.027            0                 0   
172785 2025-06-21 23:59:59.527            1                 0   

        coinc_trig_count_total  angle_tha1  angle_tha2  angle_xgse  \
0                          146    0.924455    0.999994    0.045184   
1                          110    0.924456    0.999994    0.045184   
2                          110    0.924456    0.999994    0.045184   
3                          110    0.924456    0.999994    0.045184   
4                          110    0.924456    0.999994    0.045184   
...                        ...         ...         ...         ...   
172781                     444    0.924390    0.999995    0.041773   
172782                     444    0.924390    0.999995    0.041773   
172783                     444    0.924390    0.999995    0.041773   
172784                     444    0.924390    0.999995    0.041773   
172785                     444    0.924390    0.999995    0.041773   

        angle_ygse  angle_zgse  peak_det_counts  rej_counts      obs_time  \
0        89.979738   89.959613                0           0  1.750464e+09   
1        89.979738   89.959613                1           0  1.750464e+09   
2        89.979738   89.959613                1           0  1.750464e+09   
3        89.979738   89.959613                0           0  1.750464e+09   
4        89.979738   89.959613                0           0  1.750464e+09   
...            ...         ...              ...         ...           ...   
172781   89.989551   89.959555                1           0  1.750550e+09   
172782   89.989551   89.959555                0           0  1.750550e+09   
172783   89.989551   89.959555                0           0  1.750550e+09   
172784   89.989551   89.959555                0           0  1.750550e+09   
172785   89.989551   89.959555                0           0  1.750550e+09   

        spacecraft_xpos  spacecraft_xvel  spacecraft_ypos  spacecraft_yvel  \
0          1.273305e+06         0.009543    -53310.220341         0.297323   
1          1.273305e+06         0.009543    -53310.071679         0.297323   
2          1.273305e+06         0.009543    -53309.923018         0.297323   
3          1.273305e+06         0.009543    -53309.774356         0.297323   
4          1.273305e+06         0.009543    -53309.625665         0.297323   
...                 ...              ...              ...              ...   
172781     1.274103e+06         0.008838    -27494.483756         0.300187   
172782     1.274103e+06         0.008838    -27494.333662         0.300187   
172783     1.274103e+06         0.008838    -27494.183569         0.300187   
172784     1.274103e+06         0.008838    -27494.033476         0.300187   
172785     1.274103e+06         0.008838    -27493.883383         0.300187   

        spacecraft_zpos  spacecraft_zvel  
0        -106259.023493        -0.002839  
1        -106259.024913        -0.002839  
2        -106259.026333        -0.002839  
3        -106259.027752        -0.002839  
4        -106259.029172        -0.002839  
...                 ...              ...  
172781   -106419.406078        -0.000880  
172782   -106419.406518        -0.000880  
172783   -106419.406958        -0.000880  
172784   -106419.407397        -0.000880  
172785   -106419.407837        -0.000880  

[172786 rows x 1

In [5]:
from src.preprocessing_cdf_file import *
dataset_df = L2_prepare_data([L2_BLK, L2_TH1,L2_TH2])

In [13]:
dataset_df

time  proton_bulk_speed  alpha_bulk_speed  \
0     2025-06-21 00:00:27.241       5.278118e+02      5.277175e+02   
1     2025-06-21 00:00:32.241       5.316419e+02      5.304577e+02   
2     2025-06-21 00:00:37.241       5.223615e+02      5.236657e+02   
3     2025-06-21 00:00:42.241       5.172195e+02      5.149521e+02   
4     2025-06-21 00:00:47.241       5.114079e+02      5.092423e+02   
...                       ...                ...               ...   
17270 2025-06-21 23:59:38.026      -1.000000e+31     -1.000000e+31   
17271 2025-06-21 23:59:43.027      -1.000000e+31     -1.000000e+31   
17272 2025-06-21 23:59:48.027      -1.000000e+31     -1.000000e+31   
17273 2025-06-21 23:59:53.027      -1.000000e+31     -1.000000e+31   
17274 2025-06-21 23:59:58.027      -1.000000e+31     -1.000000e+31   

       alpha_density  alpha_thermal  proton_density  proton_thermal  \
0       2.129289e-02   7.061646e+01    4.728935e-01    5.811940e+01   
1       2.381133e-02   7.468554e+01    5.853984e-01    6.066030e+01   
2       1.844540e-02   7.902136e+01    3.528019e-01    5.421125e+01   
3       2.230880e-02   8.429529e+01    3.733669e-01    6.022701e+01   
4       1.926183e-02   7.682361e+01    3.762946e-01    5.358589e+01   
...              ...            ...             ...             ...   
17270  -1.000000e+31  -1.000000e+31   -1.000000e+31   -1.000000e+31   
17271  -1.000000e+31  -1.000000e+31   -1.000000e+31   -1.000000e+31   
17272  -1.000000e+31  -1.000000e+31   -1.000000e+31   -1.000000e+31   
17273  -1.000000e+31  -1.000000e+31   -1.000000e+31   -1.000000e+31   
17274  -1.000000e+31  -1.000000e+31   -1.000000e+31   -1.000000e+31   

       proton_xvelocity  proton_yvelocity  proton_zvelocity  spacecraft_xpos  \
0         -5.278118e+02         40.998300         74.362111     1.273305e+06   
1         -5.316419e+02         50.764305         75.093527     1.273305e+06   
2         -5.223615e+02         60.206399         78.753785     1.273305e+06   
3         -5.172195e+02         42.802706         75.764355     1.273306e+06   
4         -5.114079e+02         41.228391         68.502695     1.273306e+06   
...                 ...               ...               ...              ...   
17270     -1.000000e+31         -4.844595         25.853663     1.274103e+06   
17271     -1.000000e+31         -5.110272         30.083116     1.274103e+06   
17272     -1.000000e+31          0.493213         35.987634     1.274103e+06   
17273     -1.000000e+31         -8.036967         25.808357     1.274103e+06   
17274     -1.000000e+31          9.656233         -5.958573     1.274103e+06   

       spacecraft_ypos  spacecraft_zpos  
0        -53303.976497   -106259.083118  
1        -53302.489909   -106259.097312  
2        -53301.003230   -106259.111507  
3        -53299.516581   -106259.125700  
4        -53298.029960   -106259.139893  
...                ...              ...  
17270    -27500.337479   -106419.388918  
17271    -27498.836489   -106419.393319  
17272    -27497.335557   -106419.397719  
17273    -27495.834595   -106419.402119  
17274    -27494.333662   -106419.406518  

[17275 rows x 13 columns]

In [ ]:
dataset_df.columns

In [ ]:
L1_AUX

## L1_AUX
---

### 🌟 Why These Variables Matter

CMEs are massive eruptions of plasma and magnetic field from the Sun. When they pass a spacecraft, they often cause:

- **Sudden increases in particle flux** (especially protons and electrons).
- **Directional changes** in particle arrival (anisotropy).
- **Coincidence triggers** across multiple detectors.
- **Temporal clustering** of high-energy events.

---

### 🔍 How the Variables Help Detect CMEs

| Variable | How It Helps | Why It's Useful |
|---------|--------------|-----------------|
| `trig_counts` | Measures triggered events across detectors and energy bins | Spikes in counts can indicate CME-driven particle bursts |
| `coin_trig_counts` / `coinc_trig_count_total` | Detects simultaneous hits across detectors | Helps confirm real events vs. noise |
| `angle_tha1`, `angle_tha2`, `angle_[x/y/z]gse` | Provides particle arrival directions | Used to trace CME origin and propagation path |
| `peak_det_counts` / `rej_counts` | Shows peak and rejected events | Helps identify saturation or filtering during intense events |
| `obs_time`, `epoch_for_cdf` | Time-stamps each measurement | Essential for correlating with solar observations or other spacecraft |
| `spacecraft_[x/y/z]pos` / `vel` | Gives spacecraft location and motion | Needed for coordinate transformation and triangulation |

---

### 🧠 Putting It All Together

In practice, you’d:
1. **Monitor `trig_counts` and `coin_trig_counts`** for sudden increases.
2. **Check `angle_tha1/tha2`** to see if particles are coming from the Sun’s direction.
3. **Use `obs_time`** to align with solar imagery (e.g., from Aditya-L1 or SOHO).
4. **Filter out false positives** using `rej_counts`, `pile_up_time`, and temperature/voltage 

In [ ]:
L1_TH1

## L1_TH1
---

### 🔍 Key zVariables for CME Analysis

| Variable | Why It Matters | How It's Used |
|---------|----------------|----------------|
| **`THA-1_spec`** | This is the **spectral data** from the THA-1 detector, likely representing energy-resolved particle counts across time, channels, and energy bins. | You analyze this to detect **sudden enhancements in particle flux**—a hallmark of CME-driven solar energetic particle (SEP) events. You can integrate over energy bins or track specific channels for spikes. |
| **`obs_time`** | Provides the **observation timestamp** for each data point. | Essential for **temporal correlation** with solar events (e.g., flares or CME onset from coronagraphs like LASCO or Aditya-L1 VELC). |
| **`epoch_for_cdf`** | Another form of time reference, often used for aligning data across instruments. | Useful for **synchronizing** THA-1 data with other spacecraft or ground-based observations. |
| **`fpga_ticks`** | Internal clock ticks from the FPGA, possibly used for **high-resolution timing**. | Can help in **fine-grained timing analysis** of particle arrival, especially if `obs_time` is coarse. |
| **`frame_num`** | Likely a sequential counter for data frames. | Helps in **data integrity checks** and aligning frames with telemetry or event logs. |

---

### 🧠 How You’d Use Them Together

1. **Extract `THA-1_spec`** over time and energy bins.
2. **Look for sudden increases** in counts—especially in higher energy channels.
3. **Use `obs_time` or `epoch_for_cdf`** to timestamp these spikes.
4. **Cross-reference** with CME onset times from solar imagers.
5. Optionally, use `fpga_ticks` for sub-second timing or `frame_num` to trace anomalies.

If you’d like, I can help you write a Python script to visualize `THA-1_spec` over time or detect flux thresholds. Want to give that a go?


In [ ]:
L1_TH2

In [ ]:
L2_TH1

In [ ]:
L2_TH2

In [ ]:
L2_BLK

In [ ]:
sample_df_01 = pd.DataFrame({'Time' : L2_TH1['epoch_for_cdf_mod']})
sample_df_02 = pd.DataFrame({'Time' : L2_BLK['epoch_for_cdf_mod']})

* L1_AUX[`Time`] == L1 every pairs but needs a slicing [:,0] and sample for L2 also
* L2[`Time`] != ant L1
* L2 need not slicing we can directly use value

172786

## L1_TH2
---

### 🔍 Key zVariables and Their Role

| Variable | Why It Matters | How It's Used |
|---------|----------------|----------------|
| **`THA-2_spec`** | This is the **spectral data** from the THA-2 detector—likely representing particle counts across time, channels, and energy bins. | You analyze this to detect **sudden enhancements in particle flux**, especially in higher energy bins, which are strong indicators of CME-driven solar energetic particle (SEP) events. |
| **`obs_time`** | Provides the **actual observation timestamps** for each spectral frame. | Crucial for **temporal correlation** with CME onset times from solar imagers (e.g., LASCO, Aditya-L1). |
| **`epoch_for_cdf`** | A standardized time reference used across instruments. | Helps **synchronize** THA-2 data with other spacecraft or ground-based datasets. |
| **`fpga_ticks`** | Internal FPGA clock ticks—likely high-resolution timing. | Useful for **fine-grained timing analysis** of particle arrival, especially when `obs_time` is coarse or interpolated. |
| **`frame_num`** | Sequential frame counter. | Helps ensure **data continuity** and identify any dropped or corrupted frames during CME passage. |

---

### 🧠 How You’d Use Them Together

1. **Plot `THA-2_spec` over time** using `obs_time` or `epoch_for_cdf`.
2. **Look for sudden spikes** in specific energy channels—these are your CME candidates.
3. **Use `fpga_ticks`** if you need sub-second resolution to study particle arrival timing.
4. **Validate frame integrity** with `frame_num` to ensure no data loss during the event.


## L2_TH2
---

### 🔍 CME-Relevant zVariables Explained

| Variable | Why It Matters | How It's Used |
|---------|----------------|----------------|
| **`energy_center_mod`** | Defines the **central energy values** of each spectral bin. | Sets the energy axis for interpreting flux data—essential for identifying energy-dependent particle enhancements during CME-driven SEP events. |
| **`integrated_flux_mod`**, **`integrated_flux_s15–s19_mod`** | These are **energy-resolved fluxes**, likely from different THA-2 angular sectors. | You analyze these for **sudden increases in particle flux**, especially in higher energy bins. Sector-specific data helps detect **anisotropies** in particle arrival—key for confirming solar origin. |
| **`epoch_for_cdf_mod`** | Time reference for each flux measurement. | Used to **correlate flux spikes** with CME onset times from solar imagers or other spacecraft. |
| **`spacecraft_[x/y/z]pos`** | Spacecraft’s 3D position in space. | Crucial for **transforming flux vectors** into heliocentric coordinates and for **triangulating CME propagation**. |
| **`sun_angle_tha2`** | Likely the **angle between THA-2 sectors and the Sun**. | Helps determine **directionality** of incoming particles—important for confirming that the flux spike is solar in origin. |

---

### 🧠 How You’d Use Them Together

1. **Plot `integrated_flux_s15–s19_mod` over time** using `epoch_for_cdf_mod`.
2. **Look for flux spikes** in specific energy bins—especially in higher energies.
3. **Compare across sectors** to detect directional bias (e.g., if s15–s17 show spikes but s18–s19 don’t, the CME may be coming from a specific solar longitude).
4. **Use `sun_angle_tha2`** to map sector orientation relative to the Sun.
5. **Incorporate spacecraft position** to model the CME’s trajectory or compare with other spacecraft.



## L2_BLK
---

### 🔍 CME-Relevant zVariables Explained

| Variable | Why It Matters | How It's Used |
|---------|----------------|----------------|
| **`proton_bulk_speed`** | Measures the bulk flow speed of protons. | A sudden increase (shock) followed by a plateau or decrease (ejecta) is a classic CME signature. |
| **`proton_density`** | Proton number density in the solar wind. | Spikes in density often mark the **shock front**; a drop may indicate the **magnetic cloud** or ejecta. |
| **`proton_thermal`** | Proton thermal speed or temperature. | Elevated temperature in the sheath region, followed by a drop in the ejecta, is typical of CME passage. |
| **`alpha_bulk_speed`**, **`alpha_density`**, **`alpha_thermal`** | Same as above, but for alpha particles (He²⁺). | Used to compute **alpha-to-proton ratios**, which often increase inside CMEs—an important compositional signature. |
| **`proton_[x/y/z]velocity`** | 3D velocity components of protons. | Helps identify **flow deflections** and **magnetic field orientation changes** during CME passage. |
| **`epoch_for_cdf_mod`** | Time reference for all measurements. | Essential for **temporal correlation** with solar observations and other spacecraft. |
| **`spacecraft_[x/y/z]pos`** | Spacecraft position in 3D space. | Used to **map CME trajectory** and compare with other observation points. |
| **Uncertainty variables** (e.g., `bulk_p_uncer`, `thermal_p_uncer`) | Measurement uncertainties. | Important for **error propagation** and confidence in event detection. |

---

### 🧠 How You’d Use Them Together

1. **Plot `proton_bulk_speed`, `density`, and `thermal` over time** to identify shock-sheath-ejecta structure.
2. **Look for jumps in speed and density**, followed by a drop in temperature—classic CME profile.
3. **Compare alpha and proton parameters** to detect compositional anomalies.
4. **Use velocity vectors** to analyze directional changes and possible magnetic field rotations.
5. **Correlate with `epoch_for_cdf_mod`** to match with coronagraph or EUV imagery of CME onset.


In [ ]:
sample_df_01['Time']

In [ ]:
sample_df_02['Time']

In [ ]:
(sample_df_01['Time'] == sample_df_02['Time']).sum()

In [ ]:
L1_TH1

In [ ]:
L2_BLK

In [ ]:
L1_TH2

In [ ]:
L2_TH1

In [ ]:
# Close the CDF file if it's open
import os
import pandas as pd
filter_data_into_file('cdf_data', 'cdf_data')

In [ ]:
cdf_file

In [ ]:
data_pd = pd.DataFrame({
    'DateTime' : cdf_file['epoch_for_cdf'],
    'spec' : cdf_file['THA-1_spec'],
    'obs_time' : cdf_file['obs_time'],
    'fpga_ticks' : cdf_file['fpga_ticks'],
    'frame_num' : cdf_file['frame_num']
})

data_pd

In [ ]:
L2_TH2

In [ ]:
data_pd = pd.DataFrame({
    'epoch_for_cdf': cdf_file['epoch_for_cdf'][:, 0],
    'Time': cdf_file['epoch_for_cdf'][:, 0],
    'obs_time': cdf_file['obs_time'][:, 0],
    'DC_temp': cdf_file['DC_temp_mon'][:, 0],
    'energy_mode_01': cdf_file['THA1_single_energy_mode'][:],
    'energy_mode_02': cdf_file['THA2_single_energy_mode'][:],
    'x_pos': cdf_file['spacecraft_xpos'][:],
    'y_pos': cdf_file['spacecraft_ypos'][:],
    'z_pos': cdf_file['spacecraft_zpos'][:],
    'x_vel': cdf_file['spacecraft_xvel'][:],
    'y_vel': cdf_file['spacecraft_yvel'][:],
    'z_vel': cdf_file['spacecraft_zvel'][:],
    'angle_xgse': cdf_file['angle_xgse'],
    'angle_ygse': cdf_file['angle_ygse'],
    'angle_zgse': cdf_file['angle_zgse'],
    'trig_counts': cdf_file['trig_counts'],
    'coin_trig_counts': cdf_file['coin_trig_counts'],
    'rej_counts': cdf_file['rej_counts'],
    'peak_det_counts': cdf_file['peak_det_counts'],
    'pile_up_time': cdf_file['pile_up_time'],
    'HW_LLD': cdf_file['HW_LLD'],
    'SW_LLD': cdf_file['SW_LLD'],
    'SW_ULD': cdf_file['SW_ULD']
})


In [ ]:
data_pd

In [1]:
from src.preprocessing_cdf_file import *

filter_data_into_file('C:/Users/harsh/Downloads/cdf_data_swis','cdf_data')

In [ ]:
import os
import shutil
for path in os.listdir('cdf_data/wnloads/cdf_data_swis'):
    for file in os.listdir(f'cdf_data/wnloads/cdf_data_swis/{path}'):
        shutil.move(f'cdf_data/wnloads/cdf_data_swis/{path}/{file}','C:/Users/harsh/Downloads/cdf_data_swis')


In [4]:
len('AL1_ASW91_L1_AUX_')

17

In [3]:
import pandas as pd
import shutil
import os

In [4]:
def filter_data_into_file(SRC_DIR, DEST_DIR):
    sample_df = pd.DataFrame(
        [(os.path.join(SRC_DIR, fname), fname) for fname in os.listdir(SRC_DIR)],
        columns=['file_path', 'label']
    )

    def make_folder(folder_name):
        os.makedirs(folder_name, exist_ok=True)

    def file_transfer(source, destination):
        shutil.copy(source, destination)

    # Move files into new folders
    for _, row in sample_df.iterrows():
        image = row['file_path']
        # Adjust slicing as needed for your filenames
        folder_name = image[17:-24]
        dest_folder = os.path.join(DEST_DIR, folder_name)
        make_folder(dest_folder)
        dest_path = os.path.join(dest_folder, os.path.basename(image))
        file_transfer(image, dest_path)

In [11]:
import os
import shutil

def make_folder(folder_path):
    """Creates a folder if it doesn't already exist."""
    os.makedirs(folder_path, exist_ok=True)

def file_transfer(source, destination, file_name):
    """
    Extracts folder name from file_name, creates destination folder,
    and copies the file to that folder.
    """

    # Safety: Extract folder name only if string is long enough
    if len(file_name) < 41:  # Adjust based on expected pattern
        print("Error: file_name too short for slicing. Adjust indices.")
        return

    folder_name = file_name[17:-24]
    folder_path = os.path.join(destination, folder_name)
    make_folder(folder_path)

    # Confirm file_name contains folder_name
    if folder_name in file_name:
        if not os.access(source, os.R_OK):
            print(f"Error: Cannot read source file — check permissions for {source}")
            return
        try:
            shutil.copy(source, folder_path)
            print(f"✅ File copied to: {folder_path}")
        except FileNotFoundError:
            print("❌ Error: source file not found.")
        except PermissionError:
            print("❌ Error: Permission denied. Try running the script as admin.")
        except Exception as e:
            print(f"❌ Unexpected error: {e}")
    else:
        print("⚠️ Folder name not found in file_name. File not copied.")

# Example usage
# source = r"C:/Users/harsh/Downloads/cdf_data_swis/sample.cdf"
# destination = r"C:/Users/harsh/Desktop/cdf_copies"
# file_name = "20250704T105000_ADITYA_L1_SWIS.cdf"  # Replace with actual name
# file_transfer(source, destination, file_name)

In [13]:
for file in os.listdir('C:/Users/harsh/Downloads/cdf_data_swis'):
    file_transfer(f'C:/Users/harsh/Downloads/cdf_data_swis/{file}', 'cdf_data', file)

✅ File copied to: cdf_data\20241015
✅ File copied to: cdf_data\20241017
✅ File copied to: cdf_data\20241018
✅ File copied to: cdf_data\20241019
✅ File copied to: cdf_data\20241020
✅ File copied to: cdf_data\20241021
✅ File copied to: cdf_data\20241022
✅ File copied to: cdf_data\20241023
✅ File copied to: cdf_data\20241024
✅ File copied to: cdf_data\20241025
✅ File copied to: cdf_data\20241026
✅ File copied to: cdf_data\20241027
✅ File copied to: cdf_data\20241028
✅ File copied to: cdf_data\20241029
✅ File copied to: cdf_data\20250502
✅ File copied to: cdf_data\20250504
✅ File copied to: cdf_data\20250505
✅ File copied to: cdf_data\20250506
✅ File copied to: cdf_data\20250510
✅ File copied to: cdf_data\20250511
✅ File copied to: cdf_data\20250512
✅ File copied to: cdf_data\20250513
✅ File copied to: cdf_data\20250514
✅ File copied to: cdf_data\20250515
✅ File copied to: cdf_data\20250516
✅ File copied to: cdf_data\20250517
✅ File copied to: cdf_data\20250518
✅ File copied to: cdf_data\2

In [ ]:
sample_df

## L2_TH1
---

### 🔍 CME-Relevant zVariables Explained

| Variable | Why It Matters | How It's Used |
|---------|----------------|----------------|
| **`energy_center_mod`** | Represents the **central energy values** of each spectral bin. | Defines the energy axis for interpreting flux data—essential for identifying energy-dependent enhancements during CME-driven SEP events. |
| **`integrated_flux_mod`**, **`integrated_flux_s9/s10/s11_mod`** | These are **energy-resolved flux measurements**, possibly from different sectors or detectors. | You analyze these for **sudden increases in particle flux**, especially in higher energy bins—classic CME signatures. Sector-specific fluxes help identify **anisotropies** in particle arrival. |
| **`epoch_for_cdf_mod`** | Time reference for each flux measurement. | Used to **correlate flux spikes** with CME onset times from solar imagers or other instruments. |
| **`spacecraft_[x/y/z]pos`** | Spacecraft’s position in 3D space. | Crucial for **transforming flux vectors** into heliocentric coordinates and for **triangulating CME propagation**. |
| **`sun_angle_tha1`** | Likely the **angle between the detector and the Sun** for each THA-1 channel. | Helps determine **directionality** of incoming particles—important for confirming solar origin of the event. |

---

### 🧠 How You’d Use Them Together

1. **Plot `integrated_flux_mod` over time** using `epoch_for_cdf_mod`.
2. **Identify flux spikes** in specific energy bins—especially in `s9/s10/s11` if they represent different angular sectors.
3. **Use `energy_center_mod`** to label energy bins and assess spectral hardness.
4. **Cross-reference `sun_angle_tha1`** to check if particles are arriving from the Sun’s direction.
5. **Incorporate spacecraft position** to model the CME’s trajectory or compare with other spacecraft.

